tests/db/test_tracking_operations.py

```diff
@@ -41,3 +45,97 @@ def test_set_run_status_to_killed():
+    #
+    # This is particularly important for REST clients, where
+    # TEMPORARILY_UNAVAILABLE triggers MLflow REST clients to retry the request,
+    # whereas BAD_REQUEST does not.
+    execute_counter = None
+    api_module = None
+    old_connect = None
+    old_dbapi = sqlalchemy.dialects.sqlite.pysqlite.SQLiteDialect_pysqlite.dbapi
+
+    class ConnectionWrapper:
+        """Wraps a sqlite3.Connection object."""
+
+        def __init__(self, conn):
+            self.conn = conn
+
+        def __getattr__(self, name):
+            return getattr(self.conn, name)
+
+        def cursor(self):
+            """Return a wrapped SQLite cursor."""
+            return CursorWrapper(self.conn.cursor())
+
+    class CursorWrapper:
+        """Wraps a sqlite3.Cursor object."""
+
+        def __init__(self, cursor):
+            self.cursor = cursor
+
+        def __getattr__(self, name):
+            return getattr(self.cursor, name)
+
+        def execute(self, *args, **kwargs):
+            """Wraps execute(), simulating sporadic OperationalErrors."""
+            nonlocal execute_counter
+            if execute_counter is not None and "pragma" not in args[0].lower():
+                execute_counter += 1
+                if execute_counter % 3 == 0:
+                    # Simulate a database error
+                    raise sqlite3.OperationalError("test")
+            return self.cursor.execute(*args, **kwargs)
+
+    def connect(*args, **kwargs):
+        """Wraps sqlite3.dbapi.connect(), returning a wrapped connection."""
+        global connect_counter
+        conn = old_connect(*args, **kwargs)
+        return ConnectionWrapper(conn)
+
+    def dbapi(*args, **kwargs):
+        """Wraps SQLiteDialect_pysqlite.dbapi(), returning patched dbapi."""
+        nonlocal api_module, old_connect
+        if api_module is None:
+            # Only patch the first time dbapi() is called, to avoid recursion.
+            api_module = old_dbapi(*args, **kwargs)
+            old_connect = api_module.connect
+            monkeypatch.setattr(api_module, "connect", connect)
+        return api_module
+
+    monkeypatch.setattr(sqlalchemy.dialects.sqlite.pysqlite.SQLiteDialect_pysqlite, "dbapi", dbapi)
+
+    with pytest.raises(mlflow.MlflowException):
+        with mlflow.start_run():
+            # Changing execute_counter from None to 0 enables the error
+            # simulation. We don't want to simulate errors for start_run(),
+            # because in these tests, start_run() sets up the database schema,
+            # and we don't want to simulate errors for that.
+            execute_counter = 0
+            # Perform some typical MLflow operations.
+            mlflow.log_param("p", "param")
+            mlflow.log_metric("m", 1.0)
+            mlflow.set_tag("t", "tag")
+            mlflow.pyfunc.log_model(
+                artifact_path="model", python_model=Model(), registered_model_name="model"
+            )
```

Comment:
Which line throws an exception?

Reply:
Two exceptions are thrown. This line:
```
mlflow.log_param("p", "param")
```

and this line:
```
mlflow.pyfunc.log_model(
                artifact_path="model", python_model=Model(), registered_model_name="model"
            )
```

Reply:
https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.raises might help.

<img width="872" alt="image" src="https://user-images.githubusercontent.com/17039389/200299543-a81bc9bb-fc0c-4820-a2aa-1b8b985546f6.png">


If `mlflow.log_param("p", "param")` throws an exception, the lines after that will not be executed.

Reply:
Ok, updated!